src/nodes.coffee

```diff
@@ -819,11 +816,11 @@ exports.Block = class Block extends Base
     return nodes[0] if nodes.length is 1 and nodes[0] instanceof Block
     new Block nodes
 
-  astNode: (o) ->
-    if (o.level? and o.level isnt LEVEL_TOP) and @expressions.length
-      return (new Sequence(@expressions).withLocationDataFrom @).ast o
+  ast: (o, level) ->
+    if (level? and level isnt LEVEL_TOP) and @expressions.length
+      return (new Sequence(@expressions).withLocationDataFrom @).ast o, level
```

Comment:
Some of the deep calls look for the second `level` property, so I put it back. It seems like you were trying to refactor it out, which I support, if we can do so safely. I’m not sure why it exists since it seems to always be on `o`.

Reply:
The calling interface of `ast()` is where new `level`'s get specified/passed, that's why it's the second argument to `ast()`

But once we've decoupled the *calling* interface (`ast()`) from the overrideable-method *implementation* interface (`astNode()`), we can simplify the signature of the overrideable method by implementing the logic that adds the new `level` to `o` just once in `Base::ast()`, and from then on we just need to pass around `o` (with the updated `o.level`) rather than separately pass around `o` and `level`

> Some of the deep calls look for the second `level` property

Can you point to a place in my PR where you think there's an existing issue with passing around `o`/`level`?

The pattern is that basically no node classes should override `.ast()`, so they should never need to declare a signature that accepts `level` as a separate argument

Instead they override `.astNode()`, which accepts an `o` that already has the updated `o.level`. So then if they just call eg `.ast(o)` on some child nodes, those child nodes are receiving the correct updated level (without needing to have it passed explicitly as a second argument)

Reply:
There are several nodes that override `ast`: `Block`, `Value`, `StringLiteral`, etc. I’m assuming that passing around `level` is unnecessary, but since it’s part of the signature of the parent it seemed like we _should_ have it in the overridden versions. At the very least, removing it should be its own PR, with some analysis done to confirm that it’s safe, that `o.level` always has the same value as what `level` would’ve had.